// universal-geocoder/dist/provider/mapbox/MapboxProvider.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultMapboxProviderOptions = exports.MAPBOX_GEOCODING_MODES = void 0;
var provider_1 = require("./..");
var AdminLevel_1 = __importDefault(require("../../AdminLevel"));
var MAPBOX_GEOCODING_MODES;
(function (MAPBOX_GEOCODING_MODES) {
    MAPBOX_GEOCODING_MODES["GEOCODING_MODE_PLACES"] = "mapbox.places";
    MAPBOX_GEOCODING_MODES["GEOCODING_MODE_PLACES_PERMANENT"] = "mapbox.places-permanent";
})(MAPBOX_GEOCODING_MODES = exports.MAPBOX_GEOCODING_MODES || (exports.MAPBOX_GEOCODING_MODES = {}));
exports.defaultMapboxProviderOptions = __assign(__assign({}, provider_1.defaultProviderOptions), { apiKey: "", geocodingMode: MAPBOX_GEOCODING_MODES.GEOCODING_MODE_PLACES });
var MapboxProvider = /** @class */ (function () {
    function MapboxProvider(_externalLoader, options) {
        if (options === void 0) { options = exports.defaultMapboxProviderOptions; }
        this.externalLoader = _externalLoader;
        this.options = __assign(__assign({}, exports.defaultMapboxProviderOptions), options);
        if (!this.options.apiKey) {
            throw new Error('An API key is required for the Mapbox provider. Please add it in the "apiKey" option.');
        }
    }
    MapboxProvider.prototype.geocode = function (query, callback, errorCallback) {
        var _a, _b, _c, _d, _e, _f, _g;
        var geocodeQuery = provider_1.ProviderHelpers.getGeocodeQueryFromParameter(query, provider_1.MapboxGeocodeQuery);
        if (geocodeQuery.getIp()) {
            throw new Error("The Mapbox provider does not support IP geolocation, only location geocoding.");
        }
        this.externalLoader.setOptions({
            protocol: this.options.useSsl ? "https" : "http",
            host: "api.mapbox.com",
            pathname: "geocoding/v5/" + this.options.geocodingMode + "/" + geocodeQuery.getText() + ".json",
        });
        var params = this.withCommonParams({
            bbox: geocodeQuery.getBounds()
                ? ((_a = geocodeQuery.getBounds()) === null || _a === void 0 ? void 0 : _a.west) + "," + ((_b = geocodeQuery.getBounds()) === null || _b === void 0 ? void 0 : _b.south) + "," + ((_c = geocodeQuery.getBounds()) === null || _c === void 0 ? void 0 : _c.east) + "," + ((_d = geocodeQuery.getBounds()) === null || _d === void 0 ? void 0 : _d.north)
                : undefined,
            proximity: geocodeQuery.getProximity()
                ? ((_e = geocodeQuery.getProximity()) === null || _e === void 0 ? void 0 : _e.longitude) + "," + ((_f = geocodeQuery.getProximity()) === null || _f === void 0 ? void 0 : _f.latitude)
                : undefined,
            types: geocodeQuery.getLocationTypes()
                ? (_g = geocodeQuery.getLocationTypes()) === null || _g === void 0 ? void 0 : _g.join(",") : undefined,
        }, geocodeQuery);
        this.executeRequest(params, callback, {}, {}, errorCallback);
    };
    MapboxProvider.prototype.geodecode = function (latitudeOrQuery, longitudeOrCallback, callbackOrErrorCallback, errorCallback) {
        var _a;
        var reverseQuery = provider_1.ProviderHelpers.getReverseQueryFromParameters(latitudeOrQuery, longitudeOrCallback, provider_1.MapboxReverseQuery);
        var reverseCallback = provider_1.ProviderHelpers.getCallbackFromParameters(longitudeOrCallback, callbackOrErrorCallback);
        var reverseErrorCallback = provider_1.ProviderHelpers.getErrorCallbackFromParameters(longitudeOrCallback, callbackOrErrorCallback, errorCallback);
        this.externalLoader.setOptions({
            protocol: this.options.useSsl ? "https" : "http",
            host: "api.mapbox.com",
            pathname: "geocoding/v5/" + this.options.geocodingMode + "/" + reverseQuery.getCoordinates().longitude + "," + reverseQuery.getCoordinates().latitude + ".json",
        });
        var params = this.withCommonParams({
            reverseMode: reverseQuery.getReverseMode()
                ? reverseQuery.getReverseMode()
                : undefined,
            types: reverseQuery.getLocationTypes()
                ? (_a = reverseQuery.getLocationTypes()) === null || _a === void 0 ? void 0 : _a.join(",") : "address",
        }, reverseQuery);
        this.executeRequest(params, reverseCallback, {}, {}, reverseErrorCallback);
    };
    MapboxProvider.prototype.withCommonParams = function (params, query) {
        var _a, _b;
        return __assign(__assign({}, params), { access_token: this.options.apiKey || "", country: query.getCountryCodes()
                ? (_a = query.getCountryCodes()) === null || _a === void 0 ? void 0 : _a.join(",") : (_b = this.options.countryCodes) === null || _b === void 0 ? void 0 : _b.join(","), language: query.getLocale(), limit: query.getLimit().toString() });
    };
    MapboxProvider.prototype.executeRequest = function (params, callback, headers, body, errorCallback) {
        this.externalLoader.executeRequest(params, function (data) {
            callback(data.features.map(function (result) {
                return MapboxProvider.mapToGeocoded(result);
            }));
        }, headers, body, errorCallback);
    };
    MapboxProvider.mapToGeocoded = function (result) {
        var latitude = result.geometry.coordinates[1];
        var longitude = result.geometry.coordinates[0];
        var formattedAddress = result.place_name;
        var streetNumber = result.address;
        var streetName = result.text;
        var locality;
        var postalCode;
        var region;
        var country;
        var countryCode;
        var adminLevels = [];
        var resultType = result.place_type;
        var adminLevelCode;
        (result.context || []).forEach(function (feature) {
            var type = feature.id.split(".")[0];
            switch (type) {
                case "locality":
                    locality = feature.text;
                    break;
                case "place":
                    locality = feature.text;
                    adminLevels.push(AdminLevel_1.default.create({
                        level: 2,
                        name: locality,
                    }));
                    break;
                case "postcode":
                    postalCode = feature.text;
                    break;
                case "region":
                    region = feature.text;
                    adminLevelCode = undefined;
                    if (feature.short_code && feature.short_code.match(/[A-z]{2}-/)) {
                        adminLevelCode = feature.short_code.replace(/[A-z]{2}-/, "");
                    }
                    adminLevels.push(AdminLevel_1.default.create({
                        level: 1,
                        name: region,
                        code: adminLevelCode,
                    }));
                    break;
                case "country":
                    country = feature.text;
                    countryCode = feature.short_code;
                    break;
                default:
            }
        });
        var geocoded = provider_1.MapboxGeocoded.create({
            latitude: latitude,
            longitude: longitude,
            formattedAddress: formattedAddress,
            streetNumber: streetNumber,
            streetName: streetName,
            locality: locality,
            postalCode: postalCode,
            region: region,
            adminLevels: adminLevels,
            country: country,
            countryCode: countryCode,
            resultType: resultType,
        });
        if (result.bbox) {
            geocoded = (geocoded.withBounds(result.bbox[1], result.bbox[0], result.bbox[3], result.bbox[2]));
        }
        return geocoded;
    };
    return MapboxProvider;
}());
exports.default = MapboxProvider;
