// universal-geocoder/dist/esm/provider/mapquest/MapQuestGeocoded.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import Geocoded from "../../Geocoded";
var MapQuestGeocoded = /** @class */ (function (_super) {
    __extends(MapQuestGeocoded, _super);
    function MapQuestGeocoded(_a) {
        var precision = _a.precision, precisionCode = _a.precisionCode, mapUrl = _a.mapUrl, attribution = _a.attribution, geocodedObject = __rest(_a, ["precision", "precisionCode", "mapUrl", "attribution"]);
        var _this = _super.call(this, geocodedObject) || this;
        _this.precision = precision;
        _this.precisionCode = precisionCode;
        _this.mapUrl = mapUrl;
        _this.attribution = attribution;
        return _this;
    }
    MapQuestGeocoded.create = function (object) {
        return new this(object);
    };
    MapQuestGeocoded.prototype.toObject = function () {
        return __assign(__assign({}, _super.prototype.toObject.call(this)), { precision: this.precision, precisionCode: this.precisionCode, mapUrl: this.mapUrl, attribution: this.attribution });
    };
    MapQuestGeocoded.prototype.withPrecision = function (precision) {
        return new MapQuestGeocoded(__assign(__assign({}, this.toObject()), { precision: precision }));
    };
    MapQuestGeocoded.prototype.getPrecision = function () {
        return this.precision;
    };
    MapQuestGeocoded.prototype.withPrecisionCode = function (precisionCode) {
        return new MapQuestGeocoded(__assign(__assign({}, this.toObject()), { precisionCode: precisionCode }));
    };
    MapQuestGeocoded.prototype.getPrecisionCode = function () {
        return this.precisionCode;
    };
    MapQuestGeocoded.prototype.withMapUrl = function (mapUrl) {
        return new MapQuestGeocoded(__assign(__assign({}, this.toObject()), { mapUrl: mapUrl }));
    };
    MapQuestGeocoded.prototype.getMapUrl = function () {
        return this.mapUrl;
    };
    MapQuestGeocoded.prototype.withAttribution = function (attribution) {
        return new MapQuestGeocoded(__assign(__assign({}, this.toObject()), { attribution: attribution }));
    };
    MapQuestGeocoded.prototype.getAttribution = function () {
        return this.attribution;
    };
    return MapQuestGeocoded;
}(Geocoded));
export default MapQuestGeocoded;
