// universal-geocoder/dist/provider/opencage/OpenCageGeocodeQuery.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var query_1 = require("../../query");
var OpenCageGeocodeQuery = /** @class */ (function (_super) {
    __extends(OpenCageGeocodeQuery, _super);
    function OpenCageGeocodeQuery(_a) {
        var countryCodes = _a.countryCodes, proximity = _a.proximity, minConfidence = _a.minConfidence, noRecord = _a.noRecord, geocodeQueryObject = __rest(_a, ["countryCodes", "proximity", "minConfidence", "noRecord"]);
        var _this = _super.call(this, geocodeQueryObject) || this;
        _this.countryCodes = countryCodes;
        _this.proximity = proximity;
        if (minConfidence &&
            (minConfidence.toString() !==
                parseInt(minConfidence.toString(), 10).toString() ||
                minConfidence < 1 ||
                minConfidence > 10)) {
            throw new Error('The "minConfidence" parameter must be an integer from 1 to 10.');
        }
        _this.minConfidence = minConfidence;
        _this.noRecord = noRecord;
        return _this;
    }
    OpenCageGeocodeQuery.create = function (object) {
        return new this(object);
    };
    OpenCageGeocodeQuery.prototype.toObject = function () {
        return __assign(__assign({}, _super.prototype.toObject.call(this)), { countryCodes: this.countryCodes, proximity: this.proximity, minConfidence: this.minConfidence, noRecord: this.noRecord });
    };
    OpenCageGeocodeQuery.prototype.withCountryCodes = function (countryCodes) {
        return new OpenCageGeocodeQuery(__assign(__assign({}, this.toObject()), { countryCodes: countryCodes }));
    };
    OpenCageGeocodeQuery.prototype.getCountryCodes = function () {
        return this.countryCodes;
    };
    OpenCageGeocodeQuery.prototype.withProximity = function (proximity) {
        return new OpenCageGeocodeQuery(__assign(__assign({}, this.toObject()), { proximity: proximity }));
    };
    OpenCageGeocodeQuery.prototype.getProximity = function () {
        return this.proximity;
    };
    OpenCageGeocodeQuery.prototype.withMinConfidence = function (minConfidence) {
        return new OpenCageGeocodeQuery(__assign(__assign({}, this.toObject()), { minConfidence: minConfidence }));
    };
    OpenCageGeocodeQuery.prototype.getMinConfidence = function () {
        return this.minConfidence;
    };
    OpenCageGeocodeQuery.prototype.withNoRecord = function (noRecord) {
        return new OpenCageGeocodeQuery(__assign(__assign({}, this.toObject()), { noRecord: noRecord }));
    };
    OpenCageGeocodeQuery.prototype.getNoRecord = function () {
        return this.noRecord;
    };
    return OpenCageGeocodeQuery;
}(query_1.GeocodeQuery));
exports.default = OpenCageGeocodeQuery;
