// universal-geocoder/dist/esm/provider/mapbox/MapboxGeocodeQuery.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { GeocodeQuery } from "../../query";
var MapboxGeocodeQuery = /** @class */ (function (_super) {
    __extends(MapboxGeocodeQuery, _super);
    function MapboxGeocodeQuery(_a) {
        var countryCodes = _a.countryCodes, proximity = _a.proximity, types = _a.types, fuzzyMatch = _a.fuzzyMatch, geocodeQueryObject = __rest(_a, ["countryCodes", "proximity", "types", "fuzzyMatch"]);
        var _this = _super.call(this, geocodeQueryObject) || this;
        _this.countryCodes = countryCodes;
        if (proximity && (!proximity.latitude || !proximity.longitude)) {
            throw new Error('The "proximity" parameter must be an object with the keys: "latitude", "longitude".');
        }
        _this.proximity = proximity;
        _this.types = types;
        _this.fuzzyMatch = fuzzyMatch;
        return _this;
    }
    MapboxGeocodeQuery.create = function (object) {
        return new this(object);
    };
    MapboxGeocodeQuery.prototype.toObject = function () {
        return __assign(__assign({}, _super.prototype.toObject.call(this)), { countryCodes: this.countryCodes, proximity: this.proximity, types: this.types, fuzzyMatch: this.fuzzyMatch });
    };
    MapboxGeocodeQuery.prototype.withCountryCodes = function (countryCodes) {
        return new MapboxGeocodeQuery(__assign(__assign({}, this.toObject()), { countryCodes: countryCodes }));
    };
    MapboxGeocodeQuery.prototype.getCountryCodes = function () {
        return this.countryCodes;
    };
    MapboxGeocodeQuery.prototype.withProximity = function (proximity) {
        return new MapboxGeocodeQuery(__assign(__assign({}, this.toObject()), { proximity: proximity }));
    };
    MapboxGeocodeQuery.prototype.getProximity = function () {
        return this.proximity;
    };
    MapboxGeocodeQuery.prototype.withTypes = function (types) {
        return new MapboxGeocodeQuery(__assign(__assign({}, this.toObject()), { types: types }));
    };
    MapboxGeocodeQuery.prototype.getTypes = function () {
        return this.types;
    };
    MapboxGeocodeQuery.prototype.withFuzzyMatch = function (fuzzyMatch) {
        return new MapboxGeocodeQuery(__assign(__assign({}, this.toObject()), { fuzzyMatch: fuzzyMatch }));
    };
    MapboxGeocodeQuery.prototype.getFuzzyMatch = function () {
        return this.fuzzyMatch;
    };
    return MapboxGeocodeQuery;
}(GeocodeQuery));
export default MapboxGeocodeQuery;
