// universal-geocoder/dist/cjs/provider/googlemaps/GoogleMapsReverseQuery.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var query_1 = require("../../query");
var GoogleMapsReverseQuery = /** @class */ (function (_super) {
    __extends(GoogleMapsReverseQuery, _super);
    function GoogleMapsReverseQuery(_a) {
        var types = _a.types, precisions = _a.precisions, channel = _a.channel, reverseQueryObject = __rest(_a, ["types", "precisions", "channel"]);
        var _this = _super.call(this, reverseQueryObject) || this;
        _this.types = types;
        _this.precisions = precisions;
        _this.channel = channel;
        return _this;
    }
    GoogleMapsReverseQuery.create = function (object) {
        return new this(object);
    };
    GoogleMapsReverseQuery.prototype.toObject = function () {
        return __assign(__assign({}, _super.prototype.toObject.call(this)), { types: this.types, precisions: this.precisions, channel: this.channel });
    };
    GoogleMapsReverseQuery.prototype.withTypes = function (types) {
        return new GoogleMapsReverseQuery(__assign(__assign({}, this.toObject()), { types: types }));
    };
    GoogleMapsReverseQuery.prototype.getTypes = function () {
        return this.types;
    };
    GoogleMapsReverseQuery.prototype.withPrecisions = function (precisions) {
        return new GoogleMapsReverseQuery(__assign(__assign({}, this.toObject()), { precisions: precisions }));
    };
    GoogleMapsReverseQuery.prototype.getPrecisions = function () {
        return this.precisions;
    };
    GoogleMapsReverseQuery.prototype.withChannel = function (channel) {
        return new GoogleMapsReverseQuery(__assign(__assign({}, this.toObject()), { channel: channel }));
    };
    GoogleMapsReverseQuery.prototype.getChannel = function () {
        return this.channel;
    };
    return GoogleMapsReverseQuery;
}(query_1.ReverseQuery));
exports.default = GoogleMapsReverseQuery;
