// universal-geocoder/dist/provider/BingProvider.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var provider_1 = require("./");
var Geocoded_1 = __importDefault(require("../Geocoded"));
var BingProvider = /** @class */ (function () {
    function BingProvider(_externalLoader, options) {
        if (options === void 0) { options = provider_1.defaultProviderOptions; }
        this.externalLoader = _externalLoader;
        this.options = __assign(__assign({}, provider_1.defaultProviderOptions), options);
        if (!this.options.apiKey) {
            throw new Error('An API key is required for the Bing provider. Please add it in the "apiKey" option.');
        }
    }
    BingProvider.prototype.geocode = function (query, callback, errorCallback) {
        var geocodeQuery = provider_1.ProviderHelpers.getGeocodeQueryFromParameter(query);
        if (geocodeQuery.getIp()) {
            throw new Error("The Bing provider does not support IP geolocation, only location geocoding.");
        }
        this.externalLoader.setOptions({
            protocol: this.options.useSsl ? "https" : "http",
            host: "dev.virtualearth.net",
            pathname: "REST/v1/Locations/" + geocodeQuery.getText(),
        });
        var params = {
            key: this.options.apiKey,
            jsonpCallback: this.options.useJsonp ? "jsonp" : undefined,
        };
        this.executeRequest(params, callback, {}, {}, errorCallback);
    };
    BingProvider.prototype.geodecode = function (latitudeOrQuery, longitudeOrCallback, callbackOrErrorCallback, errorCallback) {
        var reverseQuery = provider_1.ProviderHelpers.getReverseQueryFromParameters(latitudeOrQuery, longitudeOrCallback);
        var reverseCallback = provider_1.ProviderHelpers.getCallbackFromParameters(longitudeOrCallback, callbackOrErrorCallback);
        var reverseErrorCallback = provider_1.ProviderHelpers.getErrorCallbackFromParameters(longitudeOrCallback, callbackOrErrorCallback, errorCallback);
        this.externalLoader.setOptions({
            protocol: this.options.useSsl ? "https" : "http",
            host: "dev.virtualearth.net",
            pathname: "REST/v1/Locations/" + reverseQuery.getCoordinates().latitude + "," + reverseQuery.getCoordinates().longitude,
        });
        var params = {
            key: this.options.apiKey,
            jsonpCallback: this.options.useJsonp ? "jsonp" : undefined,
        };
        this.executeRequest(params, reverseCallback, {}, {}, reverseErrorCallback);
    };
    BingProvider.prototype.executeRequest = function (params, callback, headers, body, errorCallback) {
        this.externalLoader.executeRequest(params, function (data) {
            callback(data.resourceSets[0].resources.map(function (result) {
                return BingProvider.mapToGeocoded(result);
            }));
        }, headers, body, errorCallback);
    };
    BingProvider.mapToGeocoded = function (result) {
        var latitude = result.point.coordinates[0];
        var longitude = result.point.coordinates[1];
        var formattedAddress = result.address.formattedAddress;
        var streetName = result.address.addressLine;
        var _a = result.address, locality = _a.locality, postalCode = _a.postalCode;
        var region = result.address.adminDistrict;
        var country = result.address.countryRegion;
        var geocoded = Geocoded_1.default.create({
            latitude: latitude,
            longitude: longitude,
            formattedAddress: formattedAddress,
            streetName: streetName,
            locality: locality,
            postalCode: postalCode,
            region: region,
            country: country,
        });
        geocoded = geocoded.withBounds(result.bbox[0], result.bbox[1], result.bbox[2], result.bbox[3]);
        return geocoded;
    };
    return BingProvider;
}());
exports.default = BingProvider;
