// universal-geocoder/dist/cjs/provider/nominatim/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenStreetMapReverseQuery = exports.NominatimReverseQuery = exports.defaultOpenStreetMapProviderOptions = exports.OpenStreetMapProvider = exports.NominatimProvider = exports.OpenStreetMapGeocodeQuery = exports.NominatimGeocodeQuery = exports.OpenStreetMapGeocoded = exports.NominatimGeocoded = void 0;
var NominatimGeocoded_1 = require("./NominatimGeocoded");
Object.defineProperty(exports, "NominatimGeocoded", { enumerable: true, get: function () { return __importDefault(NominatimGeocoded_1).default; } });
var NominatimGeocoded_2 = require("./NominatimGeocoded");
Object.defineProperty(exports, "OpenStreetMapGeocoded", { enumerable: true, get: function () { return __importDefault(NominatimGeocoded_2).default; } });
__exportStar(require("./NominatimGeocoded"), exports);
var NominatimGeocodeQuery_1 = require("./NominatimGeocodeQuery");
Object.defineProperty(exports, "NominatimGeocodeQuery", { enumerable: true, get: function () { return __importDefault(NominatimGeocodeQuery_1).default; } });
var NominatimGeocodeQuery_2 = require("./NominatimGeocodeQuery");
Object.defineProperty(exports, "OpenStreetMapGeocodeQuery", { enumerable: true, get: function () { return __importDefault(NominatimGeocodeQuery_2).default; } });
__exportStar(require("./NominatimGeocodeQuery"), exports);
var NominatimProvider_1 = require("./NominatimProvider");
Object.defineProperty(exports, "NominatimProvider", { enumerable: true, get: function () { return __importDefault(NominatimProvider_1).default; } });
var NominatimProvider_2 = require("./NominatimProvider");
Object.defineProperty(exports, "OpenStreetMapProvider", { enumerable: true, get: function () { return __importDefault(NominatimProvider_2).default; } });
__exportStar(require("./NominatimProvider"), exports);
var NominatimProvider_3 = require("./NominatimProvider");
Object.defineProperty(exports, "defaultOpenStreetMapProviderOptions", { enumerable: true, get: function () { return NominatimProvider_3.defaultNominatimProviderOptions; } });
var NominatimReverseQuery_1 = require("./NominatimReverseQuery");
Object.defineProperty(exports, "NominatimReverseQuery", { enumerable: true, get: function () { return __importDefault(NominatimReverseQuery_1).default; } });
var NominatimReverseQuery_2 = require("./NominatimReverseQuery");
Object.defineProperty(exports, "OpenStreetMapReverseQuery", { enumerable: true, get: function () { return __importDefault(NominatimReverseQuery_2).default; } });
__exportStar(require("./NominatimReverseQuery"), exports);
