// universal-geocoder/dist/cjs/provider/nominatim/NominatimReverseQuery.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var query_1 = require("../../query");
var NominatimReverseQuery = /** @class */ (function (_super) {
    __extends(NominatimReverseQuery, _super);
    function NominatimReverseQuery(_a) {
        var zoom = _a.zoom, shape = _a.shape, shapeThreshold = _a.shapeThreshold, reverseQueryObject = __rest(_a, ["zoom", "shape", "shapeThreshold"]);
        var _this = _super.call(this, reverseQueryObject) || this;
        _this.zoom = zoom;
        if (shape && !["geojson", "kml", "svg", "text"].includes(shape)) {
            throw new Error('The "shape" parameter can only have the following values: "geojson", "kml", "svg", "text".');
        }
        _this.shape = shape;
        _this.shapeThreshold = shapeThreshold;
        return _this;
    }
    NominatimReverseQuery.create = function (object) {
        return new this(object);
    };
    NominatimReverseQuery.prototype.toObject = function () {
        return __assign(__assign({}, _super.prototype.toObject.call(this)), { zoom: this.zoom, shape: this.shape, shapeThreshold: this.shapeThreshold });
    };
    NominatimReverseQuery.prototype.withZoom = function (zoom) {
        return new NominatimReverseQuery(__assign(__assign({}, this.toObject()), { zoom: zoom }));
    };
    NominatimReverseQuery.prototype.getZoom = function () {
        return this.zoom;
    };
    NominatimReverseQuery.prototype.withShape = function (shape) {
        return new NominatimReverseQuery(__assign(__assign({}, this.toObject()), { shape: shape }));
    };
    NominatimReverseQuery.prototype.getShape = function () {
        return this.shape;
    };
    NominatimReverseQuery.prototype.withShapeThreshold = function (shapeThreshold) {
        return new NominatimReverseQuery(__assign(__assign({}, this.toObject()), { shapeThreshold: shapeThreshold }));
    };
    NominatimReverseQuery.prototype.getShapeThreshold = function () {
        return this.shapeThreshold;
    };
    return NominatimReverseQuery;
}(query_1.ReverseQuery));
exports.default = NominatimReverseQuery;
