// universal-geocoder/dist/cjs/provider/nominatim/NominatimGeocodeQuery.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var query_1 = require("../../query");
var NominatimGeocodeQuery = /** @class */ (function (_super) {
    __extends(NominatimGeocodeQuery, _super);
    function NominatimGeocodeQuery(_a) {
        var countryCodes = _a.countryCodes, excludePlaceIds = _a.excludePlaceIds, bounded = _a.bounded, bounds = _a.bounds, dedupe = _a.dedupe, shape = _a.shape, shapeThreshold = _a.shapeThreshold, geocodeQueryObject = __rest(_a, ["countryCodes", "excludePlaceIds", "bounded", "bounds", "dedupe", "shape", "shapeThreshold"]);
        var _this = _super.call(this, __assign({ bounds: bounds }, geocodeQueryObject)) || this;
        _this.countryCodes = countryCodes;
        _this.excludePlaceIds = excludePlaceIds;
        if (bounded && !bounds) {
            throw new Error('The "bounded" parameter can only be used with the "bounds" parameter.');
        }
        _this.bounded = bounded;
        _this.dedupe = dedupe;
        if (shape && !["geojson", "kml", "svg", "text"].includes(shape)) {
            throw new Error('The "shape" parameter can only have the following values: "geojson", "kml", "svg", "text".');
        }
        _this.shape = shape;
        _this.shapeThreshold = shapeThreshold;
        return _this;
    }
    NominatimGeocodeQuery.create = function (object) {
        return new this(object);
    };
    NominatimGeocodeQuery.prototype.toObject = function () {
        return __assign(__assign({}, _super.prototype.toObject.call(this)), { countryCodes: this.countryCodes, excludePlaceIds: this.excludePlaceIds, bounded: this.bounded, dedupe: this.dedupe, shape: this.shape, shapeThreshold: this.shapeThreshold });
    };
    NominatimGeocodeQuery.prototype.withCountryCodes = function (countryCodes) {
        return new NominatimGeocodeQuery(__assign(__assign({}, this.toObject()), { countryCodes: countryCodes }));
    };
    NominatimGeocodeQuery.prototype.getCountryCodes = function () {
        return this.countryCodes;
    };
    NominatimGeocodeQuery.prototype.withExcludePlaceIds = function (excludePlaceIds) {
        return new NominatimGeocodeQuery(__assign(__assign({}, this.toObject()), { excludePlaceIds: excludePlaceIds }));
    };
    NominatimGeocodeQuery.prototype.getExcludePlaceIds = function () {
        return this.excludePlaceIds;
    };
    NominatimGeocodeQuery.prototype.withBounded = function (bounded) {
        return new NominatimGeocodeQuery(__assign(__assign({}, this.toObject()), { bounded: bounded }));
    };
    NominatimGeocodeQuery.prototype.getBounded = function () {
        return this.bounded;
    };
    NominatimGeocodeQuery.prototype.withDedupe = function (dedupe) {
        return new NominatimGeocodeQuery(__assign(__assign({}, this.toObject()), { dedupe: dedupe }));
    };
    NominatimGeocodeQuery.prototype.getDedupe = function () {
        return this.dedupe;
    };
    NominatimGeocodeQuery.prototype.withShape = function (shape) {
        return new NominatimGeocodeQuery(__assign(__assign({}, this.toObject()), { shape: shape }));
    };
    NominatimGeocodeQuery.prototype.getShape = function () {
        return this.shape;
    };
    NominatimGeocodeQuery.prototype.withShapeThreshold = function (shapeThreshold) {
        return new NominatimGeocodeQuery(__assign(__assign({}, this.toObject()), { shapeThreshold: shapeThreshold }));
    };
    NominatimGeocodeQuery.prototype.getShapeThreshold = function () {
        return this.shapeThreshold;
    };
    return NominatimGeocodeQuery;
}(query_1.GeocodeQuery));
exports.default = NominatimGeocodeQuery;
