// universal-geocoder/dist/provider/googlemaps/GoogleMapsReverseQuery.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var query_1 = require("../../query");
var GoogleMapsReverseQuery = /** @class */ (function (_super) {
    __extends(GoogleMapsReverseQuery, _super);
    function GoogleMapsReverseQuery(_a) {
        var resultTypes = _a.resultTypes, locationTypes = _a.locationTypes, channel = _a.channel, reverseQueryObject = __rest(_a, ["resultTypes", "locationTypes", "channel"]);
        var _this = _super.call(this, reverseQueryObject) || this;
        _this.resultTypes = resultTypes;
        _this.locationTypes = locationTypes;
        _this.channel = channel;
        return _this;
    }
    GoogleMapsReverseQuery.create = function (object) {
        return new this(object);
    };
    GoogleMapsReverseQuery.prototype.toObject = function () {
        return __assign(__assign({}, _super.prototype.toObject.call(this)), { resultTypes: this.resultTypes, locationTypes: this.locationTypes, channel: this.channel });
    };
    GoogleMapsReverseQuery.prototype.withResultTypes = function (resultTypes) {
        return new GoogleMapsReverseQuery(__assign(__assign({}, this.toObject()), { resultTypes: resultTypes }));
    };
    GoogleMapsReverseQuery.prototype.getResultTypes = function () {
        return this.resultTypes;
    };
    GoogleMapsReverseQuery.prototype.withLocationTypes = function (locationTypes) {
        return new GoogleMapsReverseQuery(__assign(__assign({}, this.toObject()), { locationTypes: locationTypes }));
    };
    GoogleMapsReverseQuery.prototype.getLocationTypes = function () {
        return this.locationTypes;
    };
    GoogleMapsReverseQuery.prototype.withChannel = function (channel) {
        return new GoogleMapsReverseQuery(__assign(__assign({}, this.toObject()), { channel: channel }));
    };
    GoogleMapsReverseQuery.prototype.getChannel = function () {
        return this.channel;
    };
    return GoogleMapsReverseQuery;
}(query_1.ReverseQuery));
exports.default = GoogleMapsReverseQuery;
