// universal-geocoder/dist/cjs/provider/mapquest/MapQuestLocation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MapQuestLocation = /** @class */ (function () {
    function MapQuestLocation(_a) {
        var latLng = _a.latLng, street = _a.street, adminArea1 = _a.adminArea1, country = _a.country, adminArea3 = _a.adminArea3, state = _a.state, adminArea4 = _a.adminArea4, county = _a.county, adminArea5 = _a.adminArea5, city = _a.city, postalCode = _a.postalCode, type = _a.type;
        this.latLng = latLng;
        this.street = street;
        this.adminArea1 = adminArea1;
        this.country = country;
        this.adminArea3 = adminArea3;
        this.state = state;
        this.adminArea4 = adminArea4;
        this.county = county;
        this.adminArea5 = adminArea5;
        this.city = city;
        this.postalCode = postalCode;
        this.type = type;
    }
    MapQuestLocation.create = function (object) {
        return new this(object);
    };
    MapQuestLocation.prototype.toObject = function () {
        return {
            latLng: this.latLng,
            street: this.street,
            adminArea1: this.adminArea1,
            country: this.country,
            adminArea3: this.adminArea3,
            state: this.state,
            adminArea4: this.adminArea4,
            county: this.county,
            adminArea5: this.adminArea5,
            city: this.city,
            postalCode: this.postalCode,
            type: this.type,
        };
    };
    MapQuestLocation.prototype.getLatLng = function () {
        return this.latLng;
    };
    MapQuestLocation.prototype.getStreet = function () {
        return this.street;
    };
    MapQuestLocation.prototype.getAdminArea1 = function () {
        return this.adminArea1;
    };
    MapQuestLocation.prototype.getCountry = function () {
        return this.country;
    };
    MapQuestLocation.prototype.getAdminArea3 = function () {
        return this.adminArea3;
    };
    MapQuestLocation.prototype.getState = function () {
        return this.state;
    };
    MapQuestLocation.prototype.getAdminArea4 = function () {
        return this.adminArea4;
    };
    MapQuestLocation.prototype.getCounty = function () {
        return this.county;
    };
    MapQuestLocation.prototype.getAdminArea5 = function () {
        return this.adminArea5;
    };
    MapQuestLocation.prototype.getCity = function () {
        return this.city;
    };
    MapQuestLocation.prototype.getPostalCode = function () {
        return this.postalCode;
    };
    MapQuestLocation.prototype.getType = function () {
        return this.type;
    };
    return MapQuestLocation;
}());
exports.default = MapQuestLocation;
