// universal-geocoder/dist/esm/provider/locationiq/LocationIQGeocodeQuery.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { GeocodeQuery } from "../../query";
var LocationIQGeocodeQuery = /** @class */ (function (_super) {
    __extends(LocationIQGeocodeQuery, _super);
    function LocationIQGeocodeQuery(_a) {
        var countryCodes = _a.countryCodes, excludePlaceIds = _a.excludePlaceIds, bounded = _a.bounded, bounds = _a.bounds, dedupe = _a.dedupe, geocodeQueryObject = __rest(_a, ["countryCodes", "excludePlaceIds", "bounded", "bounds", "dedupe"]);
        var _this = _super.call(this, __assign({ bounds: bounds }, geocodeQueryObject)) || this;
        _this.countryCodes = countryCodes;
        _this.excludePlaceIds = excludePlaceIds;
        if (bounded && !bounds) {
            throw new Error('The "bounded" parameter can only be used with the "bounds" parameter.');
        }
        _this.bounded = bounded;
        _this.dedupe = dedupe;
        return _this;
    }
    LocationIQGeocodeQuery.create = function (object) {
        return new this(object);
    };
    LocationIQGeocodeQuery.prototype.toObject = function () {
        return __assign(__assign({}, _super.prototype.toObject.call(this)), { countryCodes: this.countryCodes, excludePlaceIds: this.excludePlaceIds, bounded: this.bounded, dedupe: this.dedupe });
    };
    LocationIQGeocodeQuery.prototype.withCountryCodes = function (countryCodes) {
        return new LocationIQGeocodeQuery(__assign(__assign({}, this.toObject()), { countryCodes: countryCodes }));
    };
    LocationIQGeocodeQuery.prototype.getCountryCodes = function () {
        return this.countryCodes;
    };
    LocationIQGeocodeQuery.prototype.withExcludePlaceIds = function (excludePlaceIds) {
        return new LocationIQGeocodeQuery(__assign(__assign({}, this.toObject()), { excludePlaceIds: excludePlaceIds }));
    };
    LocationIQGeocodeQuery.prototype.getExcludePlaceIds = function () {
        return this.excludePlaceIds;
    };
    LocationIQGeocodeQuery.prototype.withBounded = function (bounded) {
        return new LocationIQGeocodeQuery(__assign(__assign({}, this.toObject()), { bounded: bounded }));
    };
    LocationIQGeocodeQuery.prototype.getBounded = function () {
        return this.bounded;
    };
    LocationIQGeocodeQuery.prototype.withDedupe = function (dedupe) {
        return new LocationIQGeocodeQuery(__assign(__assign({}, this.toObject()), { dedupe: dedupe }));
    };
    LocationIQGeocodeQuery.prototype.getDedupe = function () {
        return this.dedupe;
    };
    return LocationIQGeocodeQuery;
}(GeocodeQuery));
export default LocationIQGeocodeQuery;
