// universal-geocoder/dist/provider/nominatim/NominatimGeocodeQuery.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var query_1 = require("../../query");
var NominatimGeocodeQuery = /** @class */ (function (_super) {
    __extends(NominatimGeocodeQuery, _super);
    function NominatimGeocodeQuery(_a) {
        var countryCodes = _a.countryCodes, excludePlaceIds = _a.excludePlaceIds, viewBox = _a.viewBox, bounded = _a.bounded, geocodeQueryObject = __rest(_a, ["countryCodes", "excludePlaceIds", "viewBox", "bounded"]);
        var _this = _super.call(this, geocodeQueryObject) || this;
        _this.countryCodes = countryCodes;
        _this.excludePlaceIds = excludePlaceIds;
        if (viewBox && viewBox.length !== 4) {
            throw new Error('The "viewBox" parameter must be 4 numbers.');
        }
        _this.viewBox = viewBox;
        if (bounded && !viewBox) {
            throw new Error('The "bounded" parameter can only be used with the "viewBox" parameter.');
        }
        _this.bounded = bounded;
        return _this;
    }
    NominatimGeocodeQuery.create = function (object) {
        return new this(object);
    };
    NominatimGeocodeQuery.prototype.toObject = function () {
        return __assign(__assign({}, _super.prototype.toObject.call(this)), { countryCodes: this.countryCodes, excludePlaceIds: this.excludePlaceIds, viewBox: this.viewBox, bounded: this.bounded });
    };
    NominatimGeocodeQuery.prototype.withCountryCodes = function (countryCodes) {
        return new NominatimGeocodeQuery(__assign(__assign({}, this.toObject()), { countryCodes: countryCodes }));
    };
    NominatimGeocodeQuery.prototype.getCountryCodes = function () {
        return this.countryCodes;
    };
    NominatimGeocodeQuery.prototype.withExcludePlaceIds = function (excludePlaceIds) {
        return new NominatimGeocodeQuery(__assign(__assign({}, this.toObject()), { excludePlaceIds: excludePlaceIds }));
    };
    NominatimGeocodeQuery.prototype.getExcludePlaceIds = function () {
        return this.excludePlaceIds;
    };
    NominatimGeocodeQuery.prototype.withViewBox = function (viewBox) {
        return new NominatimGeocodeQuery(__assign(__assign({}, this.toObject()), { viewBox: viewBox }));
    };
    NominatimGeocodeQuery.prototype.getViewBox = function () {
        return this.viewBox;
    };
    NominatimGeocodeQuery.prototype.withBounded = function (bounded) {
        return new NominatimGeocodeQuery(__assign(__assign({}, this.toObject()), { bounded: bounded }));
    };
    NominatimGeocodeQuery.prototype.getBounded = function () {
        return this.bounded;
    };
    return NominatimGeocodeQuery;
}(query_1.GeocodeQuery));
exports.default = NominatimGeocodeQuery;
