// universal-geocoder/dist/provider/nominatim/NominatimGeocoded.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var Geocoded_1 = __importDefault(require("../../Geocoded"));
var NominatimGeocoded = /** @class */ (function (_super) {
    __extends(NominatimGeocoded, _super);
    function NominatimGeocoded(_a) {
        var displayName = _a.displayName, osmId = _a.osmId, osmType = _a.osmType, category = _a.category, type = _a.type, attribution = _a.attribution, geocodedObject = __rest(_a, ["displayName", "osmId", "osmType", "category", "type", "attribution"]);
        var _this = _super.call(this, geocodedObject) || this;
        _this.displayName = displayName;
        _this.osmId = osmId;
        _this.osmType = osmType;
        _this.category = category;
        _this.type = type;
        _this.attribution = attribution;
        return _this;
    }
    NominatimGeocoded.create = function (object) {
        return new this(object);
    };
    NominatimGeocoded.prototype.toObject = function () {
        return __assign(__assign({}, _super.prototype.toObject.call(this)), { displayName: this.displayName, osmId: this.osmId, osmType: this.osmType, category: this.category, type: this.type, attribution: this.attribution });
    };
    NominatimGeocoded.prototype.withDisplayName = function (displayName) {
        return new NominatimGeocoded(__assign(__assign({}, this.toObject()), { displayName: displayName }));
    };
    NominatimGeocoded.prototype.getDisplayName = function () {
        return this.displayName;
    };
    NominatimGeocoded.prototype.withOsmId = function (osmId) {
        return new NominatimGeocoded(__assign(__assign({}, this.toObject()), { osmId: osmId }));
    };
    NominatimGeocoded.prototype.getOsmId = function () {
        return this.osmId;
    };
    NominatimGeocoded.prototype.withOsmType = function (osmType) {
        return new NominatimGeocoded(__assign(__assign({}, this.toObject()), { osmType: osmType }));
    };
    NominatimGeocoded.prototype.getOsmType = function () {
        return this.osmType;
    };
    NominatimGeocoded.prototype.withCategory = function (category) {
        return new NominatimGeocoded(__assign(__assign({}, this.toObject()), { category: category }));
    };
    NominatimGeocoded.prototype.getCategory = function () {
        return this.category;
    };
    NominatimGeocoded.prototype.withType = function (type) {
        return new NominatimGeocoded(__assign(__assign({}, this.toObject()), { type: type }));
    };
    NominatimGeocoded.prototype.getType = function () {
        return this.type;
    };
    NominatimGeocoded.prototype.withAttribution = function (attribution) {
        return new NominatimGeocoded(__assign(__assign({}, this.toObject()), { attribution: attribution }));
    };
    NominatimGeocoded.prototype.getAttribution = function () {
        return this.attribution;
    };
    return NominatimGeocoded;
}(Geocoded_1.default));
exports.default = NominatimGeocoded;
