// universal-geocoder/dist/esm/provider/opencage/OpenCageGeocodeQuery.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { GeocodeQuery } from "../../query";
var OpenCageGeocodeQuery = /** @class */ (function (_super) {
    __extends(OpenCageGeocodeQuery, _super);
    function OpenCageGeocodeQuery(_a) {
        var countryCodes = _a.countryCodes, proximity = _a.proximity, minPrecision = _a.minPrecision, noRecord = _a.noRecord, geocodeQueryObject = __rest(_a, ["countryCodes", "proximity", "minPrecision", "noRecord"]);
        var _this = _super.call(this, geocodeQueryObject) || this;
        _this.countryCodes = countryCodes;
        if (proximity && (!proximity.latitude || !proximity.longitude)) {
            throw new Error('The "proximity" parameter must be an object with the keys: "latitude", "longitude".');
        }
        _this.proximity = proximity;
        if (minPrecision &&
            (minPrecision.toString() !==
                parseInt(minPrecision.toString(), 10).toString() ||
                minPrecision < 1 ||
                minPrecision > 10)) {
            throw new Error('The "minPrecision" parameter must be an integer from 1 to 10.');
        }
        _this.minPrecision = minPrecision;
        _this.noRecord = noRecord;
        return _this;
    }
    OpenCageGeocodeQuery.create = function (object) {
        return new this(object);
    };
    OpenCageGeocodeQuery.prototype.toObject = function () {
        return __assign(__assign({}, _super.prototype.toObject.call(this)), { countryCodes: this.countryCodes, proximity: this.proximity, minPrecision: this.minPrecision, noRecord: this.noRecord });
    };
    OpenCageGeocodeQuery.prototype.withCountryCodes = function (countryCodes) {
        return new OpenCageGeocodeQuery(__assign(__assign({}, this.toObject()), { countryCodes: countryCodes }));
    };
    OpenCageGeocodeQuery.prototype.getCountryCodes = function () {
        return this.countryCodes;
    };
    OpenCageGeocodeQuery.prototype.withProximity = function (proximity) {
        return new OpenCageGeocodeQuery(__assign(__assign({}, this.toObject()), { proximity: proximity }));
    };
    OpenCageGeocodeQuery.prototype.getProximity = function () {
        return this.proximity;
    };
    OpenCageGeocodeQuery.prototype.withMinPrecision = function (minPrecision) {
        return new OpenCageGeocodeQuery(__assign(__assign({}, this.toObject()), { minPrecision: minPrecision }));
    };
    OpenCageGeocodeQuery.prototype.getMinPrecision = function () {
        return this.minPrecision;
    };
    OpenCageGeocodeQuery.prototype.withNoRecord = function (noRecord) {
        return new OpenCageGeocodeQuery(__assign(__assign({}, this.toObject()), { noRecord: noRecord }));
    };
    OpenCageGeocodeQuery.prototype.getNoRecord = function () {
        return this.noRecord;
    };
    return OpenCageGeocodeQuery;
}(GeocodeQuery));
export default OpenCageGeocodeQuery;
