// universal-geocoder/dist/cjs/provider/yandex/YandexGeocoded.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var Geocoded_1 = __importDefault(require("../../Geocoded"));
var YandexGeocoded = /** @class */ (function (_super) {
    __extends(YandexGeocoded, _super);
    function YandexGeocoded(_a) {
        var types = _a.types, precision = _a.precision, geocodedObject = __rest(_a, ["types", "precision"]);
        var _this = _super.call(this, geocodedObject) || this;
        _this.types = types;
        _this.precision = precision;
        return _this;
    }
    YandexGeocoded.create = function (object) {
        return new this(object);
    };
    YandexGeocoded.prototype.toObject = function () {
        return __assign(__assign({}, _super.prototype.toObject.call(this)), { types: this.types, precision: this.precision });
    };
    YandexGeocoded.prototype.withTypes = function (types) {
        return new YandexGeocoded(__assign(__assign({}, this.toObject()), { types: types }));
    };
    YandexGeocoded.prototype.getTypes = function () {
        return this.types;
    };
    YandexGeocoded.prototype.withPrecision = function (precision) {
        return new YandexGeocoded(__assign(__assign({}, this.toObject()), { precision: precision }));
    };
    YandexGeocoded.prototype.getPrecision = function () {
        return this.precision;
    };
    return YandexGeocoded;
}(Geocoded_1.default));
exports.default = YandexGeocoded;
