// universal-geocoder/dist/cjs/AdminLevel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ADMIN_LEVEL_CODES = void 0;
// eslint-disable-next-line no-shadow
var ADMIN_LEVEL_CODES;
(function (ADMIN_LEVEL_CODES) {
    ADMIN_LEVEL_CODES[ADMIN_LEVEL_CODES["STATE_CODE"] = 1] = "STATE_CODE";
    ADMIN_LEVEL_CODES[ADMIN_LEVEL_CODES["COUNTY_CODE"] = 2] = "COUNTY_CODE";
})(ADMIN_LEVEL_CODES = exports.ADMIN_LEVEL_CODES || (exports.ADMIN_LEVEL_CODES = {}));
var AdminLevel = /** @class */ (function () {
    function AdminLevel(_a) {
        var level = _a.level, name = _a.name, code = _a.code;
        this.level = level;
        this.name = name;
        this.code = code;
    }
    AdminLevel.create = function (object) {
        return new this(object);
    };
    AdminLevel.prototype.toObject = function () {
        return {
            level: this.level,
            name: this.name,
            code: this.code,
        };
    };
    AdminLevel.prototype.getLevel = function () {
        return this.level;
    };
    AdminLevel.prototype.getName = function () {
        return this.name;
    };
    AdminLevel.prototype.getCode = function () {
        return this.code;
    };
    return AdminLevel;
}());
exports.default = AdminLevel;
